minetest/minetest
# Split liquid_viscosity to liquid_viscosity and move_resistance

Fixes #10736. Fixes #1822.

The goal of this PR is to completely decouple liquid flowing/spreading from liquid player movement physics.

First, this splits off the `liquid_viscosity` feature of nodes into two:

* `liquid_viscosity`: 0-7. This will continue to affect the liquid flowing speed as before, but no longer the movement speed of players.
* `move_resistance`: 0-7. This will slow down players as if they are in a liquid. 0 means no slowdown at all. If this field was not set, it will be equal to `liquid_viscosity` (for backwards-compability).

`move_resistance` also works for non-liquids (except `climbable` nodes). For non-liquids, the physics are very similar to liquids, so the slowdown applies in all directions (including up/down). The biggest difference is that you can't "swim" up or down. The "fluidity" player settings are ignored if the node is not a liquid, in this case, two hardcoded values are used for the physics (see code for details).

To keep simple, the slowdown for `move_resistance` is the same like it was for `liquid_viscosity` before. There is one exception, however: If `move_resistance` is 0, there's no slowdown. Before, the value 0 for old `liquid_viscosity` caused a very subtle bit of slowdown (barely noticable, you have to measure it to see the difference). This mini-slowdown for value 0 was removed.

Second, `liquidtype` has been decoupled to split off player movement physics from the liquid flow mechanic (these two things were affected both by `liquidtype` at once). These are the relevant fields affected.

* `liquidtype`: Determines liquidtype for liquid flowing behavior (not new), but (NEW!) does no longer affect player movement (default: `"none"`, as before)
* `liquid_move_physics`: Here you can explicitly enable or disable liquidlike-movement physics (swim up/down, slowly sink, etc.) for the node, regardless of `liquidtype`. (default: enabled if `liquidtype ~= "none"` and disabled otherwise)

## To do

Idea for future PR: `move_resistance` is currently an `u8`, but only 8 values (0-7) are actually used. For the future, support for the full range of 0-255 could be added to make it much more flexible. But this might be too much for this PR.

## How to test

Give yourself all those liquid test nodes in DevTest. In particular, get these nodes out of the Chest of Everything (if you don't have the chest: `/stuff`):

* Move-resistant node
* Move-resistant node, liquidlike
* Climbable Move-resistant node
* No-swim Liquid Source Node

Play around with the nodes. Check if everything behaves as expected and if there is no regression of the "old" liquid nodes. IMPORTANT: All old worlds should behave exactly the same, as this PR must be backwards-compatible.

Here are some illustrations to help you better understand the nodes:
* "Move resistant node" is basically a cobweb. Movement through this node is slowed in all directions, including falling.
* "Climbable Move resistant node" is as if you combine a ladder with a cobweb.
* "Move-resistant node, liquidlike" is like water when you wade through it, but it doesn't flow or spread. There are no "Source" or "Flowing" variants.
* "No-swim Liquid Source/Flowing Node" is like air when you move through it (no swimming), but it flows like water.

Also, you might want to change this setting: `movement_liquid_fluidity`. Test once with default value and one with changed value. The movement speed should always be the same for "Move-resistant node" regardless of setting, but it should change with the setting value for "Move-resistant node, liquidlike".

## Use cases
* Cobwebs
* Thick foliage
* Force fields
* Liquid oxygen (flows like water but you fall and walk through it like air)
